### ISSUE TITLE
get different version dependencies between before there is a yarn.lock file and after.

### ISSUE THREAD

**Issue by klouskingsley** [yarnpkg/yarn#3023]
**Do you want to request a *feature* or report a *bug*?**
bug

**What is the current behavior?**
I try to migrate to yarn from npm. So I test the yarn lock with such a package.json file.
``` javascript
{
  "name": "yarn0test",
  "version": "1.0.0",
  "description": "yarn test",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "dependencies": {
    "uglify-js": "2.5.0",
    "webpack": "1.14.0",
    "webpack-parallel-uglify-plugin": "0.2.0"
  },
  "author": "h",
  "license": "MIT"
}

``` 
- when I first run `yarn` to install dependencies, it generates a `yarn.lock` file. 
For uglify-js part:
``` 
uglify-js@2.5.0:
  version "2.5.0"
  resolved "http://r.cnpmjs.org/uglify-js/download/uglify-js-2.5.0.tgz#4ab5d65a4730ecb7a4fb62d3f499e2054d98fba1"
  dependencies:
    async "~0.2.6"
    source-map "~0.5.1"
    uglify-to-browserify "~1.0.0"
    yargs "~3.5.4"

uglify-js@^2.6.2:
  version "2.8.18"
  resolved "http://r.cnpmjs.org/uglify-js/download/uglify-js-2.8.18.tgz#925d14bae48ab62d1883b41afe6e2261662adb8e"
  dependencies:
    source-map "~0.5.1"
    yargs "~3.10.0"
  optionalDependencies:
    uglify-to-browserify "~1.0.0"

uglify-js@~2.7.3:
  version "2.7.5"
  resolved "http://r.cnpmjs.org/uglify-js/download/uglify-js-2.7.5.tgz#4612c0c7baaee2ba7c487de4904ae122079f2ca8"
  dependencies:
    async "~0.2.6"
    source-map "~0.5.1"
    uglify-to-browserify "~1.0.0"
    yargs "~3.10.0"
```
and the node_modules is : 
`/node_modules/uglify-js`  version: 2.5.0
`/node_modules/webpack/node_modules/uglify-js` version 2.7.5
`/node_modules/webpack-parallel-uglify-plugin/node_modules/uglify-js` version **2.8.18**

- then I remove the node_modules directory and rerun `yarn` with the yarn.lock file.
I get the node_modules like this:
`/node_modules/uglify-js` version 2.5.0
`/node_modules/webpack/node_modules/uglify-js` version 2.7.5
`/node_modules/webpack-parallel-uglify-plugin/node_modules/uglify-js` version **2.7.5**

As you see , it get different uglify-js version for webpack-parallel-uglify-plugin.

**If the current behavior is a bug, please provide the steps to reproduce.**

1, run yarn the the package.json above
2, remove node_modules dir
3, run yarn

**What is the expected behavior?**
It should get the same version before there is a yarn.lock file and after.

**Please mention your node.js, yarn and operating system version.**
node version:  4.4.7
windows 10
yarn version: 0.21.3

**Comment by bestander** [comment]
Thanks for a very good bug report!
I can reproduce it.

Looks like on a second install yarn optimises yarn.lock and removes duplicate.
It should do that on first install

**Comment by klouskingsley** [comment]
In the file `src/package-resolver.js`, I found the code
```javascript
const maxValidRange = semver.maxSatisfying(versionNumbers, range);
```

**Comment by bestander** [comment]
Thanks for more data, this needs a fix

**Comment by arcanis** [comment]
@klouskingsley We just released the 0.27.0, that should ship with a fix regarding this issue - care to try again and tell me if it's ok for your use case?

**Comment by klouskingsley** [comment]
@arcanis  It works fine in 0.27.0. 😄
